crates/utils/src/settings.rs

```diff
@@ -1,27 +1,53 @@
-use crate::location_info;
+use crate::{location_info, LemmyError};
 use anyhow::Context;
-use config::{Config, ConfigError, Environment, File};
+use deser_hjson::from_str;
+use merge::Merge;
 use serde::Deserialize;
-use std::{env, fs, io::Error, net::IpAddr, sync::RwLock};
+use std::{
+  env,
+  fs,
+  io::Error,
+  net::{IpAddr, Ipv4Addr},
+  sync::RwLock,
+};
 
-static CONFIG_FILE_DEFAULTS: &str = "config/defaults.hjson";
 static CONFIG_FILE: &str = "config/config.hjson";
 
-#[derive(Debug, Deserialize, Clone)]
+#[derive(Debug, Deserialize, Clone, Merge)]
 pub struct Settings {
   pub setup: Option<Setup>,
-  pub database: DatabaseConfig,
-  pub hostname: String,
-  pub bind: IpAddr,
-  pub port: u16,
-  pub tls_enabled: bool,
-  pub jwt_secret: String,
-  pub pictrs_url: String,
-  pub iframely_url: String,
-  pub rate_limit: RateLimitConfig,
+  pub database: Option<DatabaseConfig>,
+  pub hostname: Option<String>,
```

Comment:
In order to have an empty custom config, these all need to be `Option`s.

Reply:
I'm not happy with these `unwrap_or_default()` all over the place, its just too much boilerplate. Best alternative I can think of right now is to make these optional struct members private, and create an impl with getters, which return the correct value directly.

Reply:
Its less than ideal, and its due to the fact that we have "optional" fields that aren't *really* optional, but have defaults... things like the database config and such. But there's no way to read a config file that's missing a ton of values (with defaults defined elsewhere), unless they're options.

Reply:
Thats why I propose using getters, so the public interface is simplified.

Reply:
But how would getters be better than using rusts `Default` impl? It was made for this.

Reply:
The getter would essentially just call `unwrap_or_default()`, so the public API isnt full of useless `Option`, and call sites dont have to use `unwrap_or_default()` all over the place.

Reply:
Hrm... I'll think of a way to do this.